lib/livebook/file_system/s3.ex

```diff
@@ -26,31 +26,27 @@ defmodule Livebook.FileSystem.S3 do
   end
 
   @doc """
-  Parses file system from a configuration string.
-
-  The expected format is `"BUCKET_URL ACCESS_KEY_ID SECRET_ACCESS_KEY"`.
-
-  ## Examples
-
-      Livebook.FileSystem.S3.from_config_string("https://s3.eu-central-1.amazonaws.com/mybucket myaccesskeyid mysecret")
+  Parses file system from a configuration map.
   """
-  @spec from_config_string(String.t()) :: {:ok, t()} | {:error, String.t()}
-  def from_config_string(string) do
-    case String.split(string) do
-      [bucket_url, access_key_id, secret_access_key] ->
+  @spec from_config(map()) :: {:ok, t()} | {:error, String.t()}
+  def from_config(config) do
+    case config do
+      %{
+        bucket_url: bucket_url,
+        access_key_id: access_key_id,
+        secret_access_key: secret_access_key
+      } ->
         {:ok, new(bucket_url, access_key_id, secret_access_key)}
 
-      args ->
-        {:error, "S3 filesystem configuration expects 3 arguments, but got #{length(args)}"}
+      _config ->
+        {:error,
+         "S3 filesystem config is expected to have 3 arguments: 'bucket_url', 'access_key_id' and 'secret_access_key', but got #{inspect(config)}"}
     end
   end
 
-  @doc """
-  Formats the given file system into an equivalent configuration string.
-  """
-  @spec to_config_string(t()) :: String.t()
-  def to_config_string(file_system) do
-    "#{file_system.bucket_url} #{file_system.access_key_id} #{file_system.secret_access_key}"
+  @spec to_config(t()) :: map()
+  def to_config(%__MODULE__{} = s3) do
+    Map.from_struct(s3)
```

Comment:
I would prefer to be explicit. Match on the relevant keys and return the relevant keys. This way, if we add new fields to the struct, we don't accidentally "leak" them. :)

Reply:
Do you mean something like that? Looks a little bit strange to me and still there is a lot of room for making mistakes.
I took the `Map.take` approach but then the keys may be nil at some point...
```elixir
  @spec to_config(t()) :: map()
  def to_config(%__MODULE__{} = s3) do
    %{bucket_url: bucket_url, access_key_id: access_key_id, secret_access_key: secret_access_key} = s3

    %{bucket_url: bucket_url, access_key_id: access_key_id, secret_access_key: secret_access_key}
  end

```

Reply:
Yes, exactly. Although I have a hunch this function will no longer exist after the rewrite, so I guess we can leave it as is right now. Your call.